### ISSUE TITLE
Build error on i386 Debian build daemon

### ISSUE THREAD

**Issue by mquinson** [schnorr/pajeng#30]
Hello,

I triggered a build this night, and things did not work well on i386:
```
+ PajeSimulationException: Illegal, trace is not time-ordered in (Line: 7278, Fields: '5, Contents: '11 515.000000 12 8 14')
Test suite `/<<PKGBUILDDIR>>/tests/pj_validate_simu-mardi.tesh': NOK (</<<PKGBUILDDIR>>/tests/pj_validate_simu-mardi.tesh:34> output mismatch)


92% tests passed, 1 tests failed out of 12

Total Test time (real) =   0.44 sec

The following tests FAILED:
	 12 - pj_validate_simu-mardi (Failed)
Errors while running CTest
```

That may be related to the fact that it's a 32 bits architecture?

More info: https://buildd.debian.org/status/fetch.php?pkg=pajeng&arch=i386&ver=1.3.5-1&stamp=1579555994&raw=0

**Comment by schnorr** [comment]
The exception is triggered because one of these becomes true:
```
last->startTime() > time
last->endTime() > time
```
in PajeContainer.cc line 753 and 754. There is also another case further up in line 225 with `event->time() < endTime()`. I know that we shouldn't compare floating point values this way, and that might be instigated by the 32bit arch. But how to do these comparisons with no performance loss? These tests are executed for every single event from the input (and that might be millions). So, I've searched a bit and stumbled upon [this](https://stackoverflow.com/questions/17333/what-is-the-most-effective-way-for-float-and-double-comparison). There is no definite answer, as far as I can tell. I'm particularly fond of suggestions inspired by The art of computer programming by Knuth:
```
bool approximatelyEqual(float a, float b, float epsilon)
{
    return fabs(a - b) <= ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool essentiallyEqual(float a, float b, float epsilon)
{
    return fabs(a - b) <= ( (fabs(a) > fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool definitelyGreaterThan(float a, float b, float epsilon)
{
    return (a - b) > ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool definitelyLessThan(float a, float b, float epsilon)
{
    return (b - a) > ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}
```
Epsilon could be in the order of nanoseconds, not sure about the performance of these functions. Ideas?

**Comment by mquinson** [comment]
Hello,

the Debian package will now be dropped from the testing (and future stable) release because of that problem. We should do something :)

If you had a patch, I could already test whether it's sufficient to fix the issue in the Debian package. 

My personal opinion is that the performance is less important than the fact that it actually works, but since it seems limited to 32bits, that may be less true. Another possibility would be to forbid the construction of the package on the x86 architecture, but that would be a pitty, IMHO. 

Thanks,

**Comment by mquinson** [comment]
Hello,

the Debian package will now be dropped from the testing (and future stable) release because of that problem. We should do something :)

If you had a patch, I could already test whether it's sufficient to fix the issue in the Debian package. 

My personal opinion is that the performance is less important than the fact that it actually works, but since it seems limited to 32bits, that may be less true. Another possibility would be to forbid the construction of the package on the x86 architecture, but that would be a pitty, IMHO. 

Thanks,

**Comment by schnorr** [comment]
We'll take a look asap.

**Comment by schnorr** [comment]
Fixed with 68338cbbd722ac9256567d3b71ef3a8349564d69 and ce7bfb9b2c0e5bee13a2d55921abf289c3644ae9. There is a new release [1.3.6](https://github.com/schnorr/pajeng/releases/tag/1.3.6) (bug fixes only).

Thanks @mquinson for the report.

**Comment by mquinson** [comment]
Hello,

I finally uploaded the new version into debian, and your fix is effective. It builds on all essential architectures. There is a remaining failure on Hurd-i386, but that's not essential.
https://buildd.debian.org/status/package.php?p=pajeng

Thanks a lot,
